xwiki-platform-core/xwiki-platform-localization/xwiki-platform-localization-rest/src/main/java/org/xwiki/localization/internal/DefaultLocalizationSource.java

```diff
@@ -0,0 +1,117 @@
+import javax.inject.Named;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+import javax.ws.rs.core.Response;
+
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.component.manager.ComponentLookupException;
+import org.xwiki.component.manager.ComponentManager;
+import org.xwiki.localization.LocalizationContext;
+import org.xwiki.localization.LocalizationManager;
+import org.xwiki.localization.Translation;
+import org.xwiki.localization.rest.LocalizationSource;
+import org.xwiki.rest.XWikiRestComponent;
+
+import com.fasterxml.jackson.databind.node.JsonNodeFactory;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.xpn.xwiki.XWikiContext;
+
+/**
+ * Default implementation of {@link LocalizationSource}.
+ *
+ * @version $Id$
+ * @since 13.3RC1
+ */
+@Component
+@Named("org.xwiki.localization.internal.DefaultLocalizationSource")
+@Singleton
+public class DefaultLocalizationSource implements LocalizationSource, XWikiRestComponent
+{
+    @Inject
+    @Named("context")
+    private Provider<ComponentManager> componentManagerProvider;
+
+    @Inject
+    private Provider<XWikiContext> contextProvider;
+
+    @Inject
+    private Logger logger;
+
+    @Override
+    public Response translations(String wikiId, String locale, String prefix, List<String> keys)
+    {
+        XWikiContext xWikiContext = this.contextProvider.get();
+
+        // Save the current context wiki id. It will be restored on the method is finished.
+        String oldWikiId = xWikiContext.getWikiId();
+        try {
+            // Set the requested wiki id.
+            xWikiContext.setWikiId(wikiId);
+
+            ComponentManager componentManager = this.componentManagerProvider.get();
+            LocalizationContext localizationContext = componentManager.getInstance(LocalizationContext.class);
+            LocalizationManager localizationManager = componentManager.getInstance(LocalizationManager.class);
+
+            // Resolve the local.
+            Locale local;
+            if (StringUtils.isNotEmpty(locale)) {
+                local = Locale.forLanguageTag(locale);
+            } else {
+                local = localizationContext.getCurrentLocale();
+            }
+
+            // Computes the prefix.
+            String cleanPrefix = Objects.toString(prefix, "");
+
+            ObjectNode result = JsonNodeFactory.instance.objectNode();
+
+            // Resolve the raw translation of the requested keys.
+            for (String key : keys) {
+                String fullKey = cleanPrefix + key;
+                Translation translation = localizationManager.getTranslation(fullKey, local);
+                if (translation != null) {
+                    Object rawSource = translation.getRawSource();
+                    result.put(fullKey, String.valueOf(rawSource));
+                } else {
+                    result.putNull(fullKey);
```

Comment:
Maybe add a comment to explain why it's in the result at all.

Reply:
https://github.com/xwiki/xwiki-platform/pull/1595/files#diff-118a400f3886da870ebde91eda0d6269f44a19975059f7d189964ac02c48d623R100-R101